tiledb/sm/serialization/fragment_metadata.cc

```diff
@@ -377,35 +198,84 @@ Status fragment_metadata_from_capnp(
     // the version of a fragment is on disk, we will be serializing _on wire_ in
     // fragment_metadata_to_capnp in the "modern" (post v5) way, so we need to
     // deserialize it as well in that way.
-    frag_meta->rtree().deserialize(
-        deserializer, &domain, constants::format_version);
-  }
-
-  // It's important to do this here as init_domain depends on some fields
-  // above to be properly initialized
-  if (frag_meta_reader.hasNonEmptyDomain()) {
-    auto reader = frag_meta_reader.getNonEmptyDomain();
-    auto&& [status, ndrange] = utils::deserialize_non_empty_domain_rv(reader);
-    RETURN_NOT_OK(status);
-    // Whilst sparse gets its domain calculated, dense needs to have it
-    // set here from the deserialized data
-    if (array_schema->dense()) {
-      frag_meta->init_domain(*ndrange);
-    } else {
-      const auto& frag0_dom = *ndrange;
-      frag_meta->non_empty_domain().assign(frag0_dom.begin(), frag0_dom.end());
-    }
-  }
-
-  if (frag_meta_reader.hasGtOffsets()) {
-    generic_tile_offsets_from_capnp(
-        frag_meta_reader.getGtOffsets(), frag_meta->generic_tile_offsets());
-    loaded_metadata.footer_ = true;
-  }
-
-  frag_meta->set_loaded_metadata(loaded_metadata);
-
-  return Status::Ok();
+    rtree.deserialize(deserializer, &domain, constants::format_version);
+  }
+
+  return make_shared<FragmentMetadata>(
+      HERE(),
+      resources,
+      memory_tracker,
+      array_schema,
+      utils::capnp_list_to_vector<uint64_t>(
+          frag_meta_reader.hasFileSizes(), frag_meta_reader.getFileSizes()),
```

Comment:
I'll need a second opinion from @Shelnutt2, but I think this pattern of calling `getFileSizes()` when `hasFileSizes()` could have returned false breaks the backwards compatibility guarantees of capnp messages that are missing that field.

Reply:
Changed `capnp_list_to_vector` to accept a lambda that gets the capnp list. Capnproto might return an empty list in `get***` if `has***` returns false, but I am not entirely sure of it after reading its sources.